File: packages/python/src/bank_lint/result.py
"""Shared result type returned by every validator."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Optional

ValidationError = Literal[
    "INVALID_FORMAT",
    "INVALID_LENGTH",
    "INVALID_CHECKSUM",
    "UNSUPPORTED_COUNTRY",
]


@dataclass(frozen=True)
class ValidationResult:
    valid: bool
    value: str
    error: Optional[ValidationError] = None


def ok(value: str) -> ValidationResult:
    return ValidationResult(valid=True, value=value, error=None)


def fail(value: str, error: ValidationError) -> ValidationResult:
    return ValidationResult(valid=False, value=value, error=error)

File: packages/js/src/result.ts
/** Machine-readable reasons a value can fail validation. */
export type ValidationError =
  | "INVALID_FORMAT"
  | "INVALID_LENGTH"
  | "INVALID_CHECKSUM"
  | "UNSUPPORTED_COUNTRY";

/** The outcome of validating a bank/account value. */
export interface ValidationResult {
  valid: boolean;
  value: string;
  error: ValidationError | null;
}

export function ok(value: string): ValidationResult {
  return { valid: true, value, error: null };
}

export function fail(value: string, error: ValidationError): ValidationResult {
  return { valid: false, value, error };
}
